ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs

```diff
@@ -265,34 +272,114 @@ newResolverResource resolvConf = go
+                       timeout
+                       resolvConf
+                       rootPeersGroupsVar
+                       domainsGroupsVar = do
+  domainsGroups <- atomically $ readTVar domainsGroupsVar
+  traceWith tracer (TraceLocalRootDomains domainsGroups)
 #if !defined(mingw32_HOST_OS)
-    rr <- asyncResolverResource resolvConf
+  rr <- asyncResolverResource resolvConf
 #else
-    let rr = newResolverResource resolvConf
+  let rr = newResolverResource resolvConf
 #endif
-    withAsyncAll (map (monitorDomain rr) (NonEmpty.toList domains)) $ \asyncs ->
-      snd <$> waitAny asyncs
+  let
+      -- Flatten the local root peers groups and associate a group index to each
+      -- DomainAddress to be monitorized.
+      addresses = [ (group, address, pa)
+                  | (group, (_, m)) <- zip [0..] domainsGroups
+                  , (address, pa) <- Map.toList m ]
+      domains = [ (group, domain, pa)
+                | (group, RelayDomain domain, pa) <- addresses ]
+      ipsAddr = [ (group, IP.toSockAddr (ip, port), pa)
+                | (group, RelayAddress ip port, pa) <- addresses ]
+
+      -- Since we want to preserve the number of groups, the targets, and
+      -- the addresses within each group, we fill the TVar with
+      -- a placeholder list, in order for each monitored DomainAddress to
+      -- be updated in the correct group.
+      placeholder = map (\(t, _) -> (t, Map.empty)) domainsGroups
+      placeholder' =
+        foldr (\(group, addr, pa) ph
+                -> let (target, entry) = ph !! group
+                   in updateList group (target, Map.insert addr pa entry) ph)
+              placeholder
+              ipsAddr
+
+  atomically $
+    writeTVar rootPeersGroupsVar placeholder'
+
+  -- Launch DomainAddress monitoring threads and
+  -- wait for threads to return or for local config changes
+  withAsyncAll (map (monitorDomain rr) domains) $ \as ->
+    atomically $ do
+      if null as
+        then waitConfigChanged domainsGroups
+        else void (waitAnySTM as) <|> waitConfigChanged domainsGroups
+
+  -- Recursive call to remonitor all new/updated local root peers
+  localRootPeersProvider tracer
+                         timeout
+                         resolvConf
+                         rootPeersGroupsVar
+                         domainsGroupsVar
+
   where
-    monitorDomain :: Resource DNSorIOError DNS.Resolver -> (DomainAddress, PeerAdvertise) -> IO Void
-    monitorDomain rr0 (domain@DomainAddress {daDomain, daPortNumber}, advertisePeer) =
+    waitConfigChanged :: [(Int, Map RelayAddress PeerAdvertise)] -> STM IO ()
+    waitConfigChanged dg = do
+      dg' <- readTVar domainsGroupsVar
+      check (dg /= dg')
+
+    updateList :: Int -> a -> [a] -> [a]
+    updateList at v l =
+      case splitAt at l of
+        (before, _:after) -> before ++ v:after
+        _ -> l
+
+    resolveDomain
+      :: DNS.Resolver
+      -> (DomainAddress, PeerAdvertise)
```

Comment:
why not two arguments?

Reply:
no actual reason, I can change to 2 arguments if prefered!